gluonnlp/data/batchify/language_model.py

```diff
@@ -0,0 +1,287 @@
+
+
+class CorpusBPTTBatchify(object):
+    """Transform the dataset into batches of numericalized samples, in the way
+    that the recurrent states from last batch connects with the current batch
+    for each sample.
+
+    Each sample is of shape `(seq_len, batch_size)`. When `last_batch='keep'`, the first
+    dimension of last sample may be shorter than `seq_len`.
+
+    Parameters
+    ----------
+    vocab : gluonnlp.Vocab
+        The vocabulary to use for numericalizing the dataset. Each token will be mapped to the
+        index according to the vocabulary.
+    seq_len : int
+        The length of each of the samples for truncated back-propagation-through-time (TBPTT).
+    batch_size : int
+        The number of samples in each batch.
+    last_batch : {'keep', 'discard'}
+        How to handle the last batch if the remaining length is less than `seq_len`.
+
+        - keep: A batch with less samples than previous batches is returned. vocab.padding_token
+          is used to pad the last batch based on batch size.
+
+        - discard: The last batch is discarded if it's smaller than `(seq_len, batch_size)`.
+    """
+
+    def __init__(self,
+                 vocab,
+                 seq_len,
+                 batch_size,
+                 last_batch='keep'):
+        self._vocab = vocab
+        self._seq_len = seq_len
+        self._batch_size = batch_size
+        self._last_batch = last_batch
+        assert vocab.padding_token, 'Padding token must be specified in vocab for BPTT.'
+        self._padding_idx = vocab[vocab.padding_token]
+
+        if last_batch not in ['keep', 'discard']:
+            raise ValueError(
+                'Got invalid last_batch: "{}". Must be "keep" or "discard".'.
+                format(last_batch))
+
+        if self._last_batch == 'keep':
+            if not self._vocab.padding_token:
+                raise ValueError('vocab.padding_token must be specified '
+                                 'in vocab when last_batch="keep".')
+
+    def __call__(self, corpus):
+        if self._last_batch == 'keep':
+            coded = self._vocab[list(corpus)]
+            sample_len = math.ceil(float(len(coded)) / self._batch_size)
+            padding_size = _slice_pad_length(sample_len, self._seq_len + 1, 1) * \
+                self._batch_size + sample_len * self._batch_size - len(coded)
+            coded.extend([self._vocab[self._vocab.padding_token]] * int(padding_size))
+            assert len(coded) % self._batch_size == 0
+            assert not _slice_pad_length(len(coded) / self._batch_size, self._seq_len + 1, 1)
+        else:
+            sample_len = len(corpus) // self._batch_size
+            coded = self._vocab[corpus[:sample_len * self._batch_size]]
+        data = mx.nd.array(coded).reshape((self._batch_size, -1)).T
+        batches = slice_sequence(data, self._seq_len + 1, overlap=1)
+
+        return SimpleDataset(batches).transform(_split_data_label, lazy=False)
```

Comment:
we can probably use _LazyTransformDataset to avoid a copy

Reply:
CorpusBPTTBatchify aims at dataset which can fit into the memory. This is probably fine